### ISSUE TITLE
Adding a library breaks dotnet lambda package and also visual tools

### ISSUE THREAD

**Issue by InsidiousForce** [aws/aws-lambda-dotnet#90]
I have a Lambda function that works fine. If I deploy it to AWS with with the VS tools, it works fine. It's all fine.

Then, if I add another .NET Core library project to the solution, everything blows up. The project is a .NET Core library that I'd like to share across Lambda functions. I can target .netstandard1.6 or netstandard1.5 on the library and I get the same errors. **When using dotnet lambda package or the visual tools**, every single object from the net standard classes isn't found. System.Void, System.String etc etc.

I can add rhel.7-x64 to the csproj as a built target and it will build fine that way with the library for rhel.7-x64 just fine. I haven't tried deploying it yet as a zip but I will try to do that even though it's not the "right" way.

Any help here would be appreciated, *it breaks in this way even if I add the library project to the .sln but don't add a reference in the Lambda Function project*.

Lambda function
```

<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netcoreapp1.0</TargetFramework>
    <DebugType>portable</DebugType>
    <AssemblyName>OmegaEventHandlerService</AssemblyName>
    <OutputType>Exe</OutputType>
    <PackageId>OmegaEventHandlerService</PackageId>
    <RuntimeFrameworkVersion>1.0.4</RuntimeFrameworkVersion>
    <PackageTargetFallback>$(PackageTargetFallback);dnxcore50</PackageTargetFallback>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="System.Net.NetworkInformation" Version="4.3.0" />
    <PackageReference Include="Microsoft.Extensions.Logging" Version="1.1.1" />
    <PackageReference Include="Npgsql" Version="3.2.2" />
    <PackageReference Include="Newtonsoft.Json" Version="10.0.1" />
    <PackageReference Include="System.Text.Encodings.Web" Version="4.3.0" />
    <PackageReference Include="Amazon.Lambda.Core" Version="1.0.0" />
    <PackageReference Include="Amazon.Lambda.Serialization.Json" Version="1.0.1" />
    <PackageReference Include="Amazon.Lambda.Tools" Version="1.4.0">
      <PrivateAssets>All</PrivateAssets>
    </PackageReference>
    <PackageReference Include="Amazon.Lambda.SNSEvents" Version="1.0.0" />
    <PackageReference Include="AWSSDK.SimpleNotificationService" Version="3.3.0.10" />
  </ItemGroup>

  <ItemGroup>
    <DotNetCliToolReference Include="Amazon.Lambda.Tools" Version="1.4.0" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\OmegaCoreLib\OmegaCoreLib.csproj" />
  </ItemGroup>

</Project>
```

*It breaks even without that last ItemGroup in there, just adding the library project into the sln breaks it.*

LIbrary:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard1.6</TargetFramework>
    <AssemblyName>OmegaCoreLib</AssemblyName>
  </PropertyGroup>

  <ItemGroup>
      <PackageReference Include="Amazon.Lambda.Core" Version="1.0.0" />
      <PackageReference Include="Amazon.Lambda.Serialization.Json" Version="1.0.1" />
      <PackageReference Include="Microsoft.Extensions.Logging" Version="1.1.1" />
      <PackageReference Include="AWSSDK.SimpleNotificationService" Version="3.3.0.10" />
      <PackageReference Include="Newtonsoft.Json" Version="10.0.1" />
      <PackageReference Include="Npgsql" Version="3.2.2" />
      <PackageReference Include="AWS.Logger.AspNetCore" Version="1.0.0" />
</ItemGroup>

</Project>
```

Some output that may be useful:

```
Executing publish command
Deleted previous publish folder
... invoking 'dotnet publish', working folder 'C:\Users\kk\Code\digital-pi\OmegaEventHandlerService\bin\Release\netcoreapp1.0\publish'
... publish: Microsoft (R) Build Engine version 15.1.548.43366
... publish: Copyright (C) Microsoft Corporation. All rights reserved.
... publish:   OmegaEventHandlerService -> C:\Users\kk\Code\digital-pi\OmegaEventHandlerService\bin\Release\netcoreapp1.0\OmegaEventHandlerService.dll
... publish: C:\Users\kk\AppData\Local\Temp\.NETCoreApp,Version=v1.0.AssemblyAttributes.cs(4,20): error CS0400: The type or namespace name 'System' could not be found in the global namespace (are you missing an assembly reference?) [C:\Users\kk\Code\digital-pi\OmegaCoreLib\OmegaCoreLib.csproj]
... publish: obj\Release\netcoreapp1.0\OmegaCoreLib.AssemblyInfo.cs(6,12): error CS0246: The type or namespace name 'System' could not be found (are you missing a using directive or an assembly reference?) [C:\Users\kk\Code\digital-pi\OmegaCoreLib\OmegaCoreLib.csproj]
... publish: obj\Release\netcoreapp1.0\OmegaCoreLib.AssemblyInfo.cs(7,12): error CS0246: The type or namespace name 'System' could not be found (are you missing a using directive or an assembly reference?) [C:\Users\kk\Code\digital-pi\OmegaCoreLib\OmegaCoreLib.csproj]
```

... thousands of lines later...

```
... publish: C:\Program Files\dotnet\sdk\1.0.0\Sdks\Microsoft.NET.Sdk\build\Microsoft.NET.Sdk.targets(92,5): error : Assets file 'C:\Users\kk\Code\digital-pi\OmegaCoreLib\obj\project.assets.json' doesn't have a target for '.NETCoreApp,Version=v1.0'. Ensure you have restored this project for TargetFramework='netcoreapp1.0' and RuntimeIdentifier=''. [C:\Users\kk\Code\digital-pi\OmegaCoreLib\OmegaCoreLib.csproj]
```

It seems like it is doing something where it's not passing a RID to the build for the library.

The library accesses lambda core because it deals with lambda logging.

Anyway, this is driving me crazy and I'm not sure how to resolve it. Seems like a bug in the lambda tools as it's happening even if I don't add the reference to the library project.

**Comment by InsidiousForce** [comment]
The only way I could get this to package was to remove the library project from the solution, remove the ProjectReference from the csproj, and add:

```
<Reference Include="../OmegaCoreLib/bin/Release/netstandard1.6/OmegaCoreLib.dll" />
```

to my csjproj. Then both the Visual Studio AWS tools work and the dotnet lambda package work. This isn't ideal as I can't look into my library and work with it inside the solution, but at least it's something. I would expect that I can add a library project to my solution and not have it blow up?

**Comment by PavelSafronov** [comment]
If `dotnet publish` is not succeeding, it seems that this is an issue with Visual Studio, and not strictly related to the libraries and tools in this repo. So unfortunately we can't provide much assistance with this question.

**Comment by PavelSafronov** [comment]
Unrelated to your question: some of the packages you are depending on (System.Net.NetworkInformation, System.Text.Encodings.Web, Npgsql) are versions that depend on .NET Core 1.1, while Lambda supports .NET Core 1.0. I'm not certain why these dependencies are not causing issues for you, but they may eventually. If they do, downgrade the dependencies to .NET Core 1.0 packages. For System.* packages this is 4.1, and for Npgsql it's 3.1.10.

**Comment by InsidiousForce** [comment]
@PavelSafronov 

It's not build that's failing, it's ```dotnet lambda package``` that's failing. When running ```dotnet build``` by itself, it works fine. But that doesn't so anything for lambda packaging. When running ```dotnet lambda package``` I see the errors above in build, and the AWS Visual tools do the same exact thing.

**Comment by PavelSafronov** [comment]
`dotnet lambda package` runs `dotnet publish`, and according to the logs you provided that's the command which is failing. That's why the Lambda commands fails then.
What happens when you run `dotnet publish` on its own?

**Comment by InsidiousForce** [comment]
@PavelSafronov 

```dotnet build``` works perfectly fine. It only fails when I run ``dotnet lambda package`` - sorry this wasn't clear:

*I can add rhel.7-x64 to the csproj as a built target and it will build fine that way with the library for rhel.7-x64 just fine. I haven't tried deploying it yet as a zip but I will try to do that even though it's not the "right" way.*

I meant that if I did that I can use ```dotnet publish``` with the proper platform target and it builds fine, with the library. When I say library, I'm adding a .NET Core library (our internal business logic used by multiple lambda functions and internal command line utilities) to the same .sln that my lambda function project is in. My workaround of removing the library project from the .sln, adding a reference to the library's built  .dll, and then running ```dotnet lambda package``` works but it isn't ideal. When adding the project back to the solution and adding a reference to the .csproj to the sln, everything sees it, I can build correctly in visual studio, debug in visual studio, build and run fine with ```dotnet build``` and ```dotnet run```, I can add RIDs to the csproj for the lambda function csproj and target win10 or rhel in a ```dotnet publish``` and that works fine but ```dotnet lambda package``` throws the thousands of errors.

Hope that's clearer, apologies for the vagueness in the original post.

**Comment by InsidiousForce** [comment]
@PavelSafronov Thanks for the additional information on versioning...

I lowered the versions of System.* to 4.1 and Npgsql to 3.1.10 (these were causing no problems in production that I could see) but there is no 4.1 version of System.Text.Encodings.Web, it jumps from 4.0 to 4.3 as do many of the other packages. So I went back to 4.0.0. 

I didn't know there was any kind of correlation with these version numbers and .NET Core 1.0 vs 1.1 - they seem to all say .NET Standard 1.0 in Nuget so I assumed they'd work with .netcoreapp1.0 - but I see if I follow the chain up some of the dependencies require System.Runtime 4.3 which requires platform 1.1.0 - this stuff is at least mildly confusing to me. 

Microsoft.Extensions.Logging also says requires Runtime 4.3 and platform 1.1.0, but in production I'm using the ILambdaLogger I get passed to me versus the logging extension I use in the other parts of this project - I have a facade that sits in front of both. In the commandline projects that use the same library I log to Cloudwatch via the Microsoft.Extensions.Logging stuff and that works great, not need to do that in a Lambda function, but this allows all our code to use one framework.

Anyway I've lowered the values for the stuff I'm sure is used in the lambdas even thought it worked as-is. The build problem is still there, so it's unrelated for sure.

**Comment by PavelSafronov** [comment]
Thanks for the detailed explanation.
I've still been unable to reproduce the problem with referencing a project from my Lambda function. I've uploaded a very simple Lambda function and project combination that I was testing to [this repo](https://github.com/PavelSafronov/LambdaIssue90). If you are able to reproduce the issue with this repo and submit a pull request, this would allow us to understand what's happening and how you may be able to reference the project instead of the assembly.

**Comment by InsidiousForce** [comment]
Ok I'll see if I can reproduce my issue with your simple project. I'd say the two differences I see now are 1. that my library project isn't in the same sln folder as the lambda project, since it's shared among different slns and 2. My library has its own dependencies (I've tried adding those dependencies to the lambda csproj as well).

I will come up with the simplest set of projects that reproduces the issue

**Comment by genifycom** [comment]
You cannot use this one
<PackageReference Include="Microsoft.Extensions.Logging" Version="1.1.1" />
This is beyond the .Net Core 1.0 and will not work in Lambda. Everything must be 1.0.xxxx
Microsoft do not support their own LTS 1.0 version rules and it is VERY annoying.

**Comment by InsidiousForce** [comment]
Will lambda go to 1.1 or wait for 2.0? You're correct they are a mess at
present.

**Comment by InsidiousForce** [comment]
@PavelSafronov here is a very simple setup that shows the issue:

https://github.com/InsidiousForce/SampleHandlerService

https://github.com/InsidiousForce/SampleCoreLib

check these two out side by side. Load the .sln in SampleHandlerService or build from the command line. All good. Then use the `dotnet lambda package` or visual tools and you will see the errors I pointed out in my initial post.

Thank you

**Comment by andymac4182** [comment]
I have reproduced this issue with the `dotnet package` command. It appears when you specify a framework it attempts to build the library with that framework even though it doesn't target it.

**Comment by InsidiousForce** [comment]
I created those two projects from scratch and there is a problem with ``dotnet lambda package` but not with `dotnet publish`
<img width="963" alt="2017-04-21_14h02_16" src="https://cloud.githubusercontent.com/assets/923717/25296033/28d7feb0-269b-11e7-8606-068fc84b533c.png">
<img width="747" alt="2017-04-21_14h01_06" src="https://cloud.githubusercontent.com/assets/923717/25296037/2b66c5da-269b-11e7-865b-5cb9bf53a1fd.png">

I've tried on a separate machine I've just built and I have the same issue with `dotnet lambda package` and not with `dotnet publish`

What would I do to successfully build/package this? Remove the framework target from the library?

**Comment by thefringeninja** [comment]
Having the same problem at the moment. Seems like the solution would be for `dotnet lambda` not to force you to specify a framework, the same as the rest of the dotnet cli tooling.

**Comment by thefringeninja** [comment]
BTW, I was able to find a workaround. Any library in your solution that targets `netstandard-*` should also target `netcoreapp1.*` This way you can find any compilation issues earlier in your build pipeline when you execute `dotnet build`.

**Comment by InsidiousForce** [comment]
@thefringeninja can you detail the workaround? I add 

<TargetFramework>netstandard1.6;netcoreapp1.0</TargetFramework>

To the library, same issue.

My workaround is to run this for package:

```
dotnet msbuild /t:Restore /p:TargetFramework=netcoreapp1.0
dotnet lambda package
```

and this for deploy:

```
dotnet msbuild /t:Restore /p:TargetFramework=netcoreapp1.0
dotnet lambda deploy-function
```

For whatever reason, running that restore in the project restores both project and library correctly. In fact, once the restore command is run, the AWS visual tools work for deployment as well.

Of course I'd like to not have to have this clunky script and explain it to others, so if you can tell me more about your workaround that would be excellent.

thanks

**Comment by InsidiousForce** [comment]
OK when I set the library to .netcoreapp10 alone, I got a conflict between versions, and long story short the whole issue seems to have boiled down to the library targeting 1.0.4 and the app targeting 1.0.5, neither of which are specified in the .csproj. When I explicitly set both of them to 1.05, it seems like everything suddenly works correctly, even though I get warnings saying that I shouldn't be explicitly targeting an SDK. So it seems like doing the explicit restore with the netcoreapp1.0 in the project directory forces the library to restore that as well from the project directory. I have NO IDEA why the explicit versions used were different in the library versus the project, but I'll take the warnings since it builds. Thanks for the nudge.
